libs/ballot-interpreter-nh-next/src/layout.rs

```diff
@@ -0,0 +1,112 @@
+    let row_offset: i32 = -1;
+    let width: GridUnit = 10;
+    let height: GridUnit = 2;
+
+    let clamp_row = |row: i32| -> GridUnit {
+        return row.max(0).min(grid.geometry.grid_size.height as i32 - 1) as GridUnit;
+    };
+    let clamp_column = |column: i32| -> GridUnit {
+        return column.max(0).min(grid.geometry.grid_size.width as i32 - 1) as GridUnit;
+    };
+
+    let bubble_location = grid_position.location();
+
+    let top_left_grid_point: Point<GridUnit> = Point::new(
+        clamp_column(bubble_location.column as i32 + column_offset),
+        clamp_row(bubble_location.row as i32 + row_offset),
+    );
+    let bottom_right_grid_point: Point<GridUnit> = Point::new(
+        clamp_column(bubble_location.column as i32 + column_offset + width as i32),
+        clamp_row(bubble_location.row as i32 + row_offset + height as i32),
+    );
+
+    let top_left_subpixel_point = grid.point_for_location(
+        top_left_grid_point.x as GridUnit,
+        top_left_grid_point.y as GridUnit,
+    )?;
+    let bottom_right_subpixel_point = grid.point_for_location(
+        bottom_right_grid_point.x as GridUnit,
+        bottom_right_grid_point.y as GridUnit,
+    )?;
+    let top_left_pixel_point = Point::new(
+        top_left_subpixel_point.x as PixelPosition,
+        top_left_subpixel_point.y as PixelPosition,
+    );
+    let bottom_right_pixel_point = Point::new(
+        bottom_right_subpixel_point.x as PixelPosition,
+        bottom_right_subpixel_point.y as PixelPosition,
+    );
+
+    return Some(InterpretedContestOptionLayout {
+        option_id: grid_position.option_id().clone(),
+        bounds: Rect::from_points(top_left_pixel_point, bottom_right_pixel_point),
+    });
+}
+
+pub fn build_interpreted_page_layout(
+    grid: &TimingMarkGrid,
+    grid_layout: &GridLayout,
+    side: BallotSide,
+) -> Option<Vec<InterpretedContestLayout>> {
+    let grid_positions_by_contest = grid_layout
+        .grid_positions
+        .iter()
+        .filter(|grid_position| grid_position.location().side == side)
+        .map(|grid_position| (grid_position.contest_id(), grid_position))
+        .into_group_map();
+
+    grid_positions_by_contest
+        .iter()
+        .map(|(contest_id, grid_positions)| {
+            let options = grid_positions
+                .iter()
+                .map(|grid_position| build_option_layout(grid, grid_position))
+                .collect::<Option<Vec<_>>>()?;
+
+            // Use the union of the option bounds as an approximation of the contest bounds
+            let bounds = options
+                .iter()
+                .map(|option| option.bounds)
+                .reduce(|a, b| a.union(&b))
+                .expect("Contest must have options");
+
+            Some(InterpretedContestLayout {
+                contest_id: (*contest_id).clone(),
+                bounds,
+                options,
+            })
+        })
+        .collect::<Option<Vec<_>>>()
```

Comment:
The type parameter can be inferred.

```suggestion
        .collect()
```